lib/rgb_matrix/animation/config.ex

```diff
@@ -97,42 +156,85 @@ defmodule RGBMatrix.Animation.Config do
       iex> RGBMatrix.Animation.Config.update_config(
       ...>   hue_wave_config,
       ...>   RGBMatrix.Animation.HueWave.Config.schema(),
-      ...>   %{"direction" => "left"}
-      ...> )
+      ...>   %{"direction" => "left"})
       %RGBMatrix.Animation.HueWave.Config{direction: :left, speed: 4, width: 20}
   """
-  @spec update_config(config :: t, schema :: any, params :: map) :: t
+  @spec update_config(
+          config :: t,
+          schema :: config_schema,
+          params :: update_params
+        ) :: t
   def update_config(config, schema, params) do
-    Enum.reduce(params, config, &update_field(&1, &2, schema))
+    params
+    |> Enum.reduce(config, fn param, config ->
+      cast_and_update_field(param, config, schema)
+    end)
   end
 
-  defp create_atom_key(key) when is_binary(key), do: String.to_existing_atom(key)
-  defp create_atom_key(key) when is_atom(key), do: key
+  @spec cast_and_update_field(
+          param :: {atom | String.t(), any},
+          config :: t,
+          schema :: config_schema
+        ) :: t
+  defp cast_and_update_field({key, value} = _param, config, schema) do
+    with key <- create_atom_key(key),
+         {:ok, %type_module{} = type} <- Keyword.fetch(schema, key),
+         {:ok, value} <- type_module.cast(type, value) do
+      Map.put(config, key, value)
+    else
+      error ->
+        field_warn(error, key, value)
+        config
+    end
+  end
 
-  defp create_struct!(map, module), do: struct!(module, map)
+  @spec validate_field(schema_field, creation_params) ::
+          {atom, FieldType.field()} | nil
+  defp validate_field({key, %type_module{} = type} = _field, params) do
+    value = Map.get(params, key, type.default)
 
-  defp update_field({key, value}, config, schema) do
-    key = create_atom_key(key)
+    case type_module.validate(type, value) do
+      :ok ->
+        {key, value}
 
-    %type_module{} = type = Keyword.fetch!(schema, key)
-    {:ok, value} = type_module.cast(type, value)
-    if type_module.validate(type, value) == :error, do: value_error!(value, key)
+      error ->
+        field_warn(error, key, value)
+        {key, type.default}
+    end
+  end
+
+  @spec create_atom_key(key :: atom | String.t()) ::
+          (key :: atom) | :invalid_field
+  defp create_atom_key(key) when is_atom(key), do: key
 
-    Map.put(config, key, value)
+  defp create_atom_key(key) when is_binary(key) do
+    String.to_existing_atom(key)
+  rescue
+    ArgumentError ->
+      :invalid_field
   end
 
-  defp validate_field({key, %type_module{} = type}, params) do
-    value = Map.get(params, key, type.default)
+  @spec create_struct!(map, module) :: struct
+  defp create_struct!(map, module), do: struct!(module, map)
```

Comment:
I suppose because we're not trapping any errors from `struct!/2` that it makes more sense to do it this way.

Unless... maybe we should trap and log the errors and permanently unload that specific animation from the engine. It could fall under the scope of this PR, if you squint a little....

Reply:
Correct me if I'm wrong, but if there's an issue with creating a struct it sounds like a bug in the code, so I'd let it crash so it's easier to find the problem.